### DISCUSSION TITLE
How to test against "offline" or no network connection in browser mode?

### DISCUSSION THREAD

**Discussion by redjohnsh** [vitest-dev/vitest#6847] [Q&A]
I'm building this http client that maps some fetch errors to more readable errors. For this, I need to test against lack of internet connection. I know that playwright has some "offline" mode and I want to use it with vitest. Any suggestion?

In the same way, I need to test against different network error such as cors, dns, etc.

**Comment by hi-ogawa** [comment]
> I know that playwright has some "offline" mode and I want to use it with vitest. 

You can access playwright's `Page` or `BrowserContext` API from custom commands https://vitest.dev/guide/browser/commands.html#custom-commands

Calling `context.setOffline` https://playwright.dev/docs/api/class-browsercontext#browser-context-set-offline is technically possible, but that might have an issue since it might also block how underlying Vitest runtime communicates with the server.

**Reply by jamsinclair** [reply]
Thanks! This seems to work, although the types complain.

I came up with this:
```ts
import { BrowserCommand } from 'vitest/node'

export const setOffline: BrowserCommand<[boolean]> = async (
  ctx,
  isOffline: boolean,
) => {
  //@ts-expect-error Vitest types don't seem to be correct or bad configuration on my side
  ctx.context.setOffline(isOffline)
}
```

Then in your Vitest browser config:
```ts
{
   \\ ...
  browser: {
    provider: 'playwright',
    commands: {
      setOffline,
    }
  }
}
```

Then to use it in your tests:
```ts
import { commands } from '@vitest/browser/context'

commands.setOffline(true)
```
